File: source/cgi_files/cgi_script/file.py
#!/usr/bin/env python3
import os

# -- query function --------------#
def query_var(query_string):
    if (query_string):
        query_params = query_string.split('&')
        params_dict = {}
        for param in query_params:
            if "=" in param:
                name, value = param.split('=')
                params_dict[name] = value
        return params_dict
    return None
#* -- end query function  --------#



#  -- template function ----------#
def get_page(params_dict):
    if (params_dict):
        title = params_dict.get('title', 'ff')
        heading = params_dict.get('heading', 'ff')
        message = params_dict.get('message', 'ff')
    if (os.environ.get('VALID_COOKIE') == "NO"):
        with open('./public/assets/no_cookie.html', 'r') as f:
            template = f.read()
        html = template
    elif (params_dict and len(params_dict) >= 3):
        with open('public/assets/template.html', 'r') as f:
            template = f.read()
        html = template.format(title=title, heading=heading, message=message)
    else:
        with open('public/assets/post_template.html', 'r') as f:
            template = f.read()
        html = template

    return html
#* -- end template function ------#



# ---- start env ---------------#
# CONTENT_TYPE = os.environ.get('CONTENT_TYPE', 'text/plain')
query_string = os.environ.get('QUERY_STRING', '')
body_string = os.environ.get('REQUEST_BODY', '')
request_method = os.environ.get('REQUEST_METHOD')
#* --- end env -----------------#

query_data = query_var(query_string)
body_data = query_var(body_string)

if (query_data):
    print(get_page(query_data))
else:
    print(get_page(body_data))

File: source/cgi_files/cgi_script/file.go
package main

import (
    "fmt"
    "os"
    "strings"
)

// -- query function --------------//
func queryVar(queryString string) map[string]string {
    if queryString != "" {
        queryParams := strings.Split(queryString, "&")
        paramsDict := make(map[string]string)
        for _, param := range queryParams {
            parts := strings.Split(param, "=")
            paramsDict[parts[0]] = parts[1]
        }
        return paramsDict
    }
    return nil
}
// * -- end query function  --------//


//  -- template function ----------//
func getPage(paramsDict map[string]string) string {
    var html string
    var title string
    var heading string
    var message string
    if paramsDict != nil {
        title = paramsDict["title"]
        heading = paramsDict["heading"]
        message = paramsDict["message"]
    }
    if os.Getenv("VALID_COOKIE") == "NO" {
        template, err := os.ReadFile("./public/assets/no_cookie.html")
        if err != nil {
            fmt.Println(err)
            return ""
        }
        html = string(template)
    } else if len(paramsDict) >= 3 {
        template, err := os.ReadFile("public/assets/template.html")
        if err != nil {
            fmt.Println(err)
            return ""
        }
        html = string(template)
        html = strings.ReplaceAll(html, "{title}", title)
        html = strings.ReplaceAll(html, "{heading}", heading)
        html = strings.ReplaceAll(html, "{message}", message)
    } else {
        template, err := os.ReadFile("public/assets/post_template.html")
        if err != nil {
            fmt.Println(err)
            return ""
        }
        html = string(template)
    }
    return html
}

// * -- end template function ------//



func main() {
    contentType := os.Getenv("CONTENT_TYPE")
    if contentType == "" {
            contentType = "text/plain"
        }
    queryString := os.Getenv("QUERY_STRING")
    bodyString := os.Getenv("REQUEST_BODY")
    requestMethod := os.Getenv("REQUEST_METHOD")
    queryData := queryVar(queryString)
    bodyData := queryVar(bodyString)
    
    var html string
    if requestMethod == "GET" {
        html = getPage(queryData)
    } else {
        html = getPage(bodyData)
    }
    fmt.Println(html)
    }
